homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,234 @@
+DEFAULT_TIMEOUT = 2
+DEFAULT_ZONE2 = False
+DEFAULT_ZONE3 = False
+
+CONFIG_SCHEMA = vol.Schema({vol.Optional(CONF_HOST): str})
+
+SETTINGS_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): vol.All(
+            int, vol.Range(min=1)
+        ),
+        vol.Optional(CONF_SHOW_ALL_SOURCES, default=DEFAULT_SHOW_SOURCES): bool,
+        vol.Optional(CONF_ZONE2, default=DEFAULT_ZONE2): bool,
+        vol.Optional(CONF_ZONE3, default=DEFAULT_ZONE3): bool,
+    }
+)
+
+
+class DenonAvrFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Denon AVR config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the Denon AVR flow."""
+        self.host = None
+        self.serial_number = None
+        self.model_name = None
+        self.timeout = DEFAULT_TIMEOUT
+        self.show_all_sources = DEFAULT_SHOW_SOURCES
+        self.zone2 = DEFAULT_ZONE2
+        self.zone3 = DEFAULT_ZONE3
+        self.d_receivers = []
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            # check if IP address is set manually
+            host = user_input.get(CONF_HOST)
+            if host:
+                self.host = host
+                return await self.async_step_settings()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
+            # More than one receiver could be discovered by that method
+            if len(self.d_receivers) == 1:
+                self.host = self.d_receivers[0]["host"]
+                return await self.async_step_settings()
+            if len(self.d_receivers) > 1:
+                # show selection form
+                return await self.async_step_select()
+
+            errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple receivers found."""
+        errors = {}
+        if user_input is not None:
+            self.host = user_input["select_host"]
+            return await self.async_step_settings()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_host"): vol.In(
+                    [d_receiver["host"] for d_receiver in self.d_receivers]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_settings(self, user_input=None):
+        """Allow the user to specify settings."""
+        if user_input is not None:
+            # Get config option that have defaults
+            self.timeout = user_input[CONF_TIMEOUT]
```

Comment:
Unless there is a strong case for adjustable timeouts, as developers we should pick a reasonable number here and not expose this as an option.  The user shouldn't have to know enough about how things work to make the decision.

Reply:
well it was in the yaml configuration and I know there are some users that got problems with the default timeout of 2 seconds, espacially on old receiver models.
But for the majority 2 seconds is just fine, but it is nice to have it as an option if you are experiancing problems.

Reply:
Is there any reason to not set it to 5 seconds?  That should cover all the use cases it seems.

Reply:
I could set it to 5 seconds yes, that schould not be a problem I think.
Just do not want to upset users that are now able to configure the timeout and will not be if I remove it...

Reply:
Do you want me to remove it?

Reply:
Lets set it to 5 for now and take it out, if there is a use case presented that isn't covered by 5s, it can be added back to the options flow later.

Reply:
I have removed the timeout option from the config flow and hard coded it to 5 seconds